Modules/Loadable/Markups/MRML/vtkMRMLMarkupsCurveNode.cxx

```diff
@@ -183,6 +189,205 @@ double vtkMRMLMarkupsCurveNode::GetCurveLengthBetweenStartEndPointsWorld(vtkIdTy
+	this->SetControlPointLabelsWorld(originalLabels, originalControlPoints);
+}
+
+//---------------------------------------------------------------------------
+vtkIdType vtkMRMLMarkupsCurveNode::GetClosestControlPointIndexToPositionWorld(const double posWorld[3])
+{
+	vtkNew<vtkPoints> controlPoints;
+	this->GetControlPointPositionsWorld(controlPoints);
+
+	if (!controlPoints)
+	{
+		return -1;
+	}
+
+	vtkIdType closestControlPointID = 0;
+	vtkIdType numberOfPoints = controlPoints->GetNumberOfPoints();
+	double closestPointDistance = vtkMath::Distance2BetweenPoints(posWorld, controlPoints->GetPoint(0));
+
+	for (vtkIdType pointIndex = 1; pointIndex < numberOfPoints; pointIndex++)
+	{
+		double* nextPoint = controlPoints->GetPoint(pointIndex);
+		double nextPointDistance = vtkMath::Distance2BetweenPoints(posWorld, nextPoint);
+		if (nextPointDistance < closestPointDistance)
+		{
+			closestPointDistance = nextPointDistance;
+			closestControlPointID = pointIndex;
+		}
+	}
+
+	return closestControlPointID;
+}
+//---------------------------------------------------------------------------
+bool vtkMRMLMarkupsCurveNode::ConstrainPointsToSurface(vtkPoints* originalPoints, vtkPoints* normalVectors, vtkPolyData* surfacePolydata, vtkPoints* surfacePoints)
+{
+	if (originalPoints->GetNumberOfPoints()!= normalVectors->GetNumberOfPoints())
+	{
+		vtkGenericWarningMacro("vtkMRMLMarkupsCurveNode::ResamplePoints failed: invalid inputs");
+		return false;
+	}
+	
+	vtkNew<vtkOBBTree> obbTree;
+	obbTree->SetDataSet(surfacePolydata);
+	obbTree->BuildLocator();
+
+	vtkNew<vtkPointLocator> pointLocator;
+	pointLocator->SetDataSet(surfacePolydata);
+	pointLocator->BuildLocator();
+
+	double originalPoint[3] = { 0.0 };
+	double rayDirection[3] = { 0.0 };
+	double exteriorPoint[3] = { 0.0 };
+
+	// Curve is expected to be close to surface. If projection distance is greater than 10% of image size
+	// it's likely due to a second external surface
+	double bounds[6] = { 0.0 };
+	surfacePolydata->GetBounds(bounds);
+	double point1[3] = { bounds[0], bounds[2], bounds[4]};
+	double point2[3] = { bounds[1], bounds[3], bounds[5] };
+	double polydataSize = vtkMath::Distance2BetweenPoints(point1,point2);
+	double rayLength = .1*sqrt(polydataSize);
+
+	for (vtkIdType controlPointIndex = 0; controlPointIndex < originalPoints->GetNumberOfPoints(); controlPointIndex++)
+	{
+		originalPoints->GetPoint(controlPointIndex, originalPoint);
+		normalVectors->GetPoint(controlPointIndex, rayDirection);
+		//cast ray and find model intersection point
+		double rayEndPoint[3] = { 0.0 };
+		rayEndPoint[0] = originalPoint[0] + rayDirection[0] * rayLength;
+		rayEndPoint[1] = originalPoint[1] + rayDirection[1] * rayLength;
+		rayEndPoint[2] = originalPoint[2] + rayDirection[2] * rayLength;
+
+		vtkNew<vtkIdList> intersectionIds;
+		vtkNew<vtkPoints> intersectionPoints;
+		obbTree->IntersectWithLine(originalPoint, rayEndPoint, intersectionPoints, intersectionIds);
```

Comment:
Use this method to return the first intersection: https://vtk.org/doc/nightly/html/classvtkOBBTree.html#a268ecf8a6ed5d343d5c98c386918c98b

Since the polydata does not change, we can use a locator that is probably faster: https://vtk.org/doc/nightly/html/classvtkStaticCellLocator.html#details

Reply:
The alternate method for finding the intersection requires a tolerance value to be set. It wasn't clear to me how to set this value from the documentation. I'd prefer to use the other method if it may need to change.

Reply:
Tolerance is needed for almost all geometric operations, otherwise imprecision of floating-point operations would cause lots of problems. In this specific case, the line is considered to cross a cell if it is inside the cell, or it is outside the cell but its distance from the edge of the cell is less than 'tol'.

When you use the "other" method, tolerance is defined by the [Tolerance](https://vtk.org/doc/nightly/html/classvtkLocator.html#af97088c133f28f9f6c4e617615a510e9) member variable, that's why it is not required as an input argument for the method. You can use value of Tolerance member variable.

If you have any doubts about what the two methods do, you can have a look at their source code:
- https://github.com/Kitware/VTK/blob/master/Filters/General/vtkOBBTree.cxx#L663
- https://github.com/Kitware/VTK/blob/master/Filters/General/vtkOBBTree.cxx#L896

As you can see the first method (that you used originally) does many things and we discard most of the computation results. We cannot afford to do this when we have complex meshes or when we'll compute the curve snapping from control points on-the-fly.

Reply:
Great, thanks for explaining this. I'll make this change.

Reply:
I ran into an issue with the vtkStaticCellLocator. The code compiles, but then gives the error

`The locator class - vtkStaticCellLocator does not yet support FindClosestPoint`

This only comes up rarely since it's only used when no intersection is found, but I noticed that in this case it was skipping these points rather than using the closest mesh point.